### ISSUE TITLE
Terminology improvement suggestion

### ISSUE THREAD

**Issue by dmitriz** [fantasyland/fantasy-land#217]
It seems that the same letters are used for different things if I see it right.

This is where it starts getting interesting:
https://github.com/fantasyland/fantasy-land#apply

Here `a` and `b` are the types:

```hs
ap :: Apply f => f a ~> f (a -> b) -> f b
```

But here `a` and `b` are values:

```js
a.ap(b)
```
>1. `b` must be an Apply of a function,

Further,  the `b` below corresponds to the value of type `f (a -> b)` above, not even of type `b` above, which is also present.

I've thought, using some different notation for the "lifted types" might help the reading.
Like `x` for a value of type `a`, `xy` for a value of type `f (a -> b)` and `y` for the result:

```js
y = x.ap(xy)
```

Also some quick examples may further help the understanding.
Like in this case I find the `Array` example a good illustration of
the difference between Apply and Functor:

```js
[' Sara', 'John  '].ap([v => v.trim(), v => v.length])
// => ['Sara', 'John', 5, 6] (the data wins, as in Haskell)
// or
// => ['Sara', 5, 'John', 6] (the function wins)
```

There is also further confusion with JS (and [Ramda's](http://ramdajs.com/docs/#apply)) `apply` method:

```js
// Below (a) is the tuple type, e.g. (1, 4, 5)
// ((a) -> b) ~>  [a] -> b
fn.apply(array)
```
which has arguments in the flipped order comparing to the current FL spec, 
and takes a plain function rather than a lifted one (i.e. of type `f a`).

On the other hand, [Ramda's lift](http://ramdajs.com/docs/#lift) is stating to satisfy the FL's Apply spec, but actually it takes a variadic "unlifted" function instead of a lifted unary function (i.e. `f(a -> b)`):

```js
// R.lift :: ((a) -> b) -> ([a]) -> [b]
```
Basically `R.lift` acts like Functor's `map` rather than an `ap`, if I see it right.

Let me know what you think.

**Comment by SimonRichardson** [comment]
Hi @dmitriz thanks for the issue:

> It seems that the same letters are used for different things if I see it right.

This has been discussed before to some degree, I believe the out come was something on the lines of, we're happy to change things if it improves readability, but not for the sake of it! Also I think with the inclusion of the new type signatures it should be less of a pain and the types should talk for themselves.

> Also some quick examples may further help the understanding.

I'm not sure about this, maybe if we ever split the spec up into multiple documents then I would be more of a fan.

We always except PRs :)

**Comment by dmitriz** [comment]
I've made a few notational change suggestions in this PR, and before committing to a longer rework, I'd like to see if such changes are welcome.

Further suggestion as mentioned there, includes capitalizing functors to make a clear distinction from functions. 

So this
```hs
traverse :: Applicative f, Traversable t => t a ~> (a -> f b, c -> f c) -> f (t b)
```
would become:

```hs
traverse :: Applicative F, Traversable T => T a ~> (a -> F b, c -> F c) -> F (T b)
```

or, with types also capitalized:

```hs
traverse :: Applicative F, Traversable T => T A ~> (A -> F B, C -> F C) -> F (T B)
```

Next, there is still some inconsistency -- Traversable is called T but Applicative F. I presume F stands for functor. However, both are functors. A more consistent notation would be:

```hs
traverse :: Applicative A, Traversable T ...
```

But now we have notation clash with types, so we can switch to the capitals at the end of alphabet:

```hs
traverse :: Applicative A, Traversable T => T X ~> (X -> A Y, Z -> A Z) -> F (A Y)
```

Now it is immediately clear where is the applicative (`A`) and where is the traversable (`T`)!

Further, we can re-use the letters `X, Y, Z` in their uncapitalized version to denote the corresponding values:
```hs
x :: X, y :: Y, z :: Z
```
and the functions:
```
f :: X -> A Y, g :: Z -> A Z
```
or, possibly more elegantly:
```
xy :: X -> A Y, zz :: Z -> A Z
```
which succintly emphasizes the signature.

Let me know what you think.
